pkg/p2p/kadcast/writer.go

```diff
@@ -55,26 +76,25 @@ func (w *Writer) Subscribe() {
 // Write sends a message through the Kadcast gRPC interface.
 // Depending on the value of header field, Send or Broadcast is called.
 func (w *Writer) Write(data, header []byte, priority byte) (int, error) {
-	// check header
-	if len(header) == 0 {
-		return 0, errors.New("empty message header")
+	var err error
+
+	switch {
+	case len(header) > 1:
+		// point-to-point messaging
+		err = w.writeToPoint(data, header, priority)
+	case len(header) == 1:
+		// broadcast messaging
+		err = w.writeToAll(data, header, priority)
+	default:
+		err = errors.New("empty message header")
 	}
-	// send
-	go func() {
-		var err error
-		// send a p2p message
-		if len(header) > 1 {
-			err = w.writeToPoint(data, header, priority)
-		}
-		// broadcast a message
-		if len(header) == 1 {
-			err = w.writeToAll(data, header, priority)
-		}
-		// log errors
-		if err != nil {
-			log.WithError(err).Warn("write failed")
-		}
-	}()
```

Comment:
I suppose removing the goroutine has to do with the fact that the `Consumer` already consumes the messages concurrently

Reply:
Both writeToAll/writeToPoint may block due to grpc client request they call. When this happens, runtime starts queue-ing goroutines. However our purpose with this PR is to queue messages in the Listener Ring buffer so that they are always sorted by priority/(type of topic).

If that makes sense to you @autholykos ?

Reply:
Yes it does